src/trex/molecule.py

```diff
@@ -82,3 +85,88 @@ def compute_consensus(sequences):
     bin_consens = np.argmax(matrix, axis=1)
     # convert maximum counts into consensus sequence
     return "".join(letters[bin_consens])
+
+
+def remove_too_short_clone_ids(molecules: List[Molecule], min_length: int) -> List[Molecule]:
+    """
+    Discard cloneIDs shorter than min_length
+    """
+    return [
+        m for m in molecules
+        if len(m.clone_id.replace("0", "").replace("-", "")) >= min_length
+    ]
+
+
+def correct_clone_ids_per_cell(
+    molecules: List[Molecule], max_hamming: int, min_overlap: int = 20
+) -> List[Molecule]:
+    """
+    Attempt to correct sequencing errors in the CloneID sequences of all
+    molecules looking for similar sequences in the same cell.
+    """
+    # Obtain all cloneIDs (including those with '-' and '0')
+    clone_ids = [m.clone_id for m in molecules]
+
+    # Count the full-length cloneIDs
+    counts = Counter(clone_ids)
+
+    cell_dict = defaultdict(list)
+    for molecule in molecules:
+        cell_dict[molecule.cell_id].append((molecule.umi, molecule.clone_id))
+
+    # Cluster them by Hamming distance
+    def is_similar(s, t):
+        # m = max_hamming
+        bad_chars = {'-', '0'}
+        if bad_chars & set(s) or bad_chars & set(t):
+            # Remove suffix and/or prefix where sequences do not overlap
+            s = s.lstrip("-0")
+            t = t[-len(s):]
+            t = t.lstrip("-0")
+            s = s[-len(t):]
+            s = s.rstrip("-0")
+            if len(s) < min_overlap:
+                return False
+            t = t[: len(s)]
+            t = t.rstrip("-0")
+            if len(t) < min_overlap:
+                return False
+            s = s[: len(t)]
+            # TODO allowed Hamming distance should be reduced relative to the
+            #  overlap length
+            # m = max_hamming * len(s) / len(original_length_of_s)
+        return hamming_distance(s, t) <= max_hamming
+
+    cell_correction_map = defaultdict(dict)
+
+    for cell_id, cell_molecules in cell_dict.items():
+        cell_clone_ids = set([m[1] for m in cell_molecules])
```

Comment:
Because this is now the full molecule, you can just write:
```suggestion
        cell_clone_ids = set([m.clone_id for m in cell_molecules])
```
That is easier to read because you don’t have to remember what index `1` means.